### ISSUE TITLE
False positive involving `re.sub` and `+` operator introduced in mypy 0.960

### ISSUE THREAD

**Issue by DMRobertson** [python/mypy#12961]
### Bug Report

Consider the following snippet. (This is a minified example of the problem detected in https://github.com/matrix-org/synapse/issues/12996.)

```python
import re

"#" + re.sub(
    "PATTERN",
    lambda match: match.group(0) + "ANOTHER_STRING",
    "HAYSTACK",
)
```

[Mypy 0.950 is happy with this](https://mypy-play.net/?mypy=0.950&python=3.10&flags=verbose&gist=ba60332a802eac1d760e7901624da712), but [mypy 0.960 is not](https://mypy-play.net/?mypy=0.960&python=3.10&flags=verbose&gist=a40a88d26df2613541e6402694ecedf5).

> main.py:5: error: Unsupported operand types for + ("bytes" and "str")

AFAICS this is a legitimate usage of [re.sub](https://docs.python.org/3/library/re.html?highlight=re%20sub#re.sub), consistent with [the stubs defined in typeshed.](https://github.com/python/typeshed/blob/81fd55a885e678559471710f249c7e27d02ec212/stdlib/re.pyi#L135-L146)

### Additional

Strangely, removing either of the addition operators and the corresponding string literal causes mypy 0.960 to accept the snippet. That is, applying this change

```diff
- "#" + re.sub(
+ re.sub(
```
is [accepted](https://mypy-play.net/?mypy=0.960&python=3.10&flags=verbose&gist=d08e3a3c15cad6c0914ce947fb9226fc), and applying this change

```diff
-    lambda match: match.group(0) + "ANOTHER_STRING",
+    lambda match: match.group(0),
```
to the vanilla snippet above [is also accepted](https://mypy-play.net/?mypy=0.960&python=3.10&flags=verbose&gist=8285609da597ed67ef55105862f026e3).

I'm not sure how to get mypy to reveal the type of an expression in a lambda body, but I'd guess that in the error cases it deduces `match: Match[bytes]` rather than `match: Match[str]`?

**Comment by DMRobertson** [comment]
Bisection shows that this was introduced by 7fbf4deb845d6041700d1addc6eaa2097560f0f3. I'm not sure if this is a typeshed bug or just something that a typeshed change happened to expose?

**Comment by DMRobertson** [comment]
Ahh, digging into the typeshed change (https://github.com/python/typeshed/pull/7679) this looks like this is probably a duplicate of https://github.com/python/mypy/issues/12665?

**Comment by AlexWaygood** [comment]
> Ahh, digging into the typeshed change ([python/typeshed#7679](https://github.com/python/typeshed/pull/7679)) this looks like this is probably related to (duplicate of?) #12665

Or #12773 (cc. @JukkaL, @JelleZijlstra)

**Comment by hauntsaninja** [comment]
Yup, duplicate of #12666 / 12665

**Comment by hauntsaninja** [comment]
The issue is probably something like we're not correctly silencing errors in the earlier passes for the multipass lambda logic
